File: mobile/ios/Air/SubModules/WalletCore/Stores/Activities/PendingCardMints.swift
import Foundation
import WalletContext

struct PendingCardMints: Sendable {
    enum Resolution: Equatable {
        case minted(ApiNft)
        case refunded
    }

    private var startedAtByAccountId: [String: Int64] = [:]

    mutating func recordSubmission(accountId: String, since date: Date) {
        // TON timestamps have second precision, unlike the local submission time.
        startedAtByAccountId[accountId] = Int64(date.timeIntervalSince1970.rounded(.down)) * 1_000
    }

    mutating func remove(accountId: String) {
        startedAtByAccountId[accountId] = nil
    }

    mutating func consume(accountId: String, activities: some Collection<ApiActivity>) -> Resolution? {
        guard let startedAt = startedAtByAccountId[accountId] else { return nil }
        let transactions = activities.compactMap { activity -> ApiTransactionActivity? in
            guard !activity.isLocal,
                  activity.isConfirmedOrCompleted,
                  activity.timestamp >= startedAt,
                  activity.shouldHide != true,
                  case .transaction(let transaction) = activity,
                  transaction.isIncoming,
                  transaction.type != .nftTrade,
                  getChainBySlug(transaction.slug) == .ton
            else { return nil }
            return transaction
        }
        if let nft = transactions.compactMap(\.nft).first(where: {
            $0.chain == .ton && $0.collectionAddress == MTW_CARDS_COLLECTION
        }) {
            remove(accountId: accountId)
            return .minted(nft)
        }
        if transactions.contains(where: {
            $0.fromAddress == MINT_CARD_ADDRESS && $0.comment == MINT_CARD_REFUND_COMMENT
        }) {
            remove(accountId: accountId)
            return .refunded
        }
        return nil
    }
}

File: mobile/android/air/SubModules/WalletCore/src/main/java/org/mytonwallet/app_air/walletcore/stores/PendingCardMints.kt
package org.mytonwallet.app_air.walletcore.stores

import org.mytonwallet.app_air.walletcore.MINT_CARD_ADDRESS
import org.mytonwallet.app_air.walletcore.MINT_CARD_REFUND_COMMENT
import org.mytonwallet.app_air.walletcore.MTW_CARDS_COLLECTION
import org.mytonwallet.app_air.walletcore.TONCOIN_SLUG
import org.mytonwallet.app_air.walletcore.models.blockchain.MBlockchain
import org.mytonwallet.app_air.walletcore.moshi.ApiNft
import org.mytonwallet.app_air.walletcore.moshi.ApiTransactionStatus
import org.mytonwallet.app_air.walletcore.moshi.ApiTransactionType
import org.mytonwallet.app_air.walletcore.moshi.MApiTransaction

internal class PendingCardMints {
    sealed class Resolution {
        data class Minted(val nft: ApiNft) : Resolution()
        object Refunded : Resolution()
    }

    private val startedAtByAccountId = HashMap<String, Long>()

    fun recordSubmission(accountId: String, startedAt: Long) {
        startedAtByAccountId[accountId] = startedAt / 1_000 * 1_000
    }

    fun remove(accountId: String) {
        startedAtByAccountId.remove(accountId)
    }

    fun consume(accountId: String, activities: Collection<MApiTransaction>): Resolution? {
        val startedAt = startedAtByAccountId[accountId] ?: return null
        val transactions = activities.filterIsInstance<MApiTransaction.Transaction>().filter {
            !it.isLocal() &&
                (
                    it.status == ApiTransactionStatus.CONFIRMED ||
                        it.status == ApiTransactionStatus.COMPLETED
                    ) &&
                it.timestamp >= startedAt &&
                it.shouldHide != true &&
                it.isIncoming &&
                it.type != ApiTransactionType.NFT_TRADE &&
                (it.slug == TONCOIN_SLUG || it.slug.startsWith("ton-"))
        }
        val nft = transactions.mapNotNull { it.nft }.firstOrNull {
            it.chain == MBlockchain.ton && it.collectionAddress == MTW_CARDS_COLLECTION
        }
        if (nft != null) {
            remove(accountId)
            return Resolution.Minted(nft)
        }
        if (transactions.any {
                it.fromAddress == MINT_CARD_ADDRESS && it.comment == MINT_CARD_REFUND_COMMENT
            }
        ) {
            remove(accountId)
            return Resolution.Refunded
        }
        return null
    }
}
